### PR TITLE
feat: Bitcoin chain tip validation in transaction signer and coordinator

### PR DESCRIPTION
**Pull request by netrome**
Closes: #297 

## Checklist:

- [x] I have performed a self-review of my code

### CODE REVIEW

File: signer/src/transaction_coordinator.rs [hunk 1]
```diff
@@ -344,6 +344,11 @@ where
         loop {
             let msg = self.network.receive().await?;
 
+            if msg.bitcoin_chain_tip.as_byte_array() != bitcoin_chain_tip.as_slice() {
+                tracing::warn!(msg=?msg, "concurrent wsts signing round message observed");
+                continue;
+            }
```
Comment (djordon) [inline signer/src/transaction_coordinator.rs:350]: I'm not sure that this is correct (well, it doesn't seem incorrect either). What happens if there is a bitcoin block that lands in the middle of a signing round? Wait, I think this issue is basically encapsulated by https://github.com/stacks-network/sbtc/issues/465, so nothing obvious to change here.

File: signer/src/testing/transaction_signer.rs [hunk 2]
```diff
@@ -331,20 +332,34 @@ where
...
+        .unwrap();
+
+        let coordinator_private_key = signer_info
+            .iter()
+            .find(|signer_info| {
+                PublicKey::from_private_key(&signer_info.signer_private_key)
+                    == coordinator_public_key
+            })
+            .unwrap()
+            .signer_private_key;
```
Comment (djordon) [inline signer/src/testing/transaction_signer.rs:356]: It's probably worth making a proper type that wraps `Vec<SignerInfo>` so that we can have functions for these common operations. No need to do anything in the PR though.

**Review by djordon — APPROVED** [review]
Looks good. The only substantial comment I have is related to https://github.com/stacks-network/sbtc/issues/465.

File: signer/src/transaction_signer.rs [hunk 3]
```diff
@@ -678,6 +747,19 @@ where
...
+    sender_is_coordinator: bool,
+    chain_tip_status: ChainTipStatus,
+}
+
+#[derive(Debug, Clone, Copy)]
+enum ChainTipStatus {
+    Canonical,
+    Known,
+    Unknown,
+}
```
[the remark this answers is not in the document]
Comment (netrome) [reply signer/src/transaction_signer.rs]: Sure thing, done in https://github.com/stacks-network/sbtc/pull/480/commits/be6998bddb94dcbbb911726177ac2eed0a9e2312

File: signer/src/transaction_coordinator.rs [hunk 4]
```diff
@@ -344,6 +344,11 @@ where
         loop {
             let msg = self.network.receive().await?;
 
+            if msg.bitcoin_chain_tip.as_byte_array() != bitcoin_chain_tip.as_slice() {
+                tracing::warn!(msg=?msg, "concurrent wsts signing round message observed");
```
[the remark this answers is not in the document]
Comment (netrome) [reply signer/src/transaction_coordinator.rs]: Right, fixed in https://github.com/stacks-network/sbtc/pull/480/commits/be6998bddb94dcbbb911726177ac2eed0a9e2312

File: signer/src/transaction_coordinator.rs (hunk 1 above)
Comment (netrome) [reply signer/src/transaction_coordinator.rs]: I suspect that this actually solves #465, but we should add tests for it in the context of #465 and ensure there aren't any edge cases we're missing here, and that the system works as intended. The following scenarios should be okay at least:

### Scenario 1
1. Coordinator A: Begins signing round
2. A wild bitcoin block appears
3. Coordinator B != A: Begins signing round

Coordinator A will likely not succeed because the signers would ignore it's messages as stale (although we would tolerate the scenario when it does), but Coordinator B would just ignore any messages relating to coordinator A:s signing round and should be able to continue do it's work without any problems.

Note that even though Coordinator A will hang for a while, the transaction signer task within the same binary will still be responsive to the messages of Coordinator B.

### Scenario 2
1. Coordinator A: Begins signing round
2. A wild bitcoin block appears
3. Coordinator B == A is coordinator for the new block.

Because this coordinator will likely hang and time out, we will have a delay in processing this block. But once the timeout is reached, the coordinator should be able to continue coordinating signing rounds as usual.

Comment (netrome) [reply signer/src/transaction_coordinator.rs]: ### Scenario three
1. Two competing forks exist but not all blocks are relayed to all signers, some signers follow one fork other signers follow another fork.
2. Coordinator A and B simultaneously try to coordinate a signing round.

Since they ignore messages from other block hashes, they should not interfere with each other here which I think is the correct behavior in this scenario. This could lead to a temporary stalemate if the partition is somewhat equal between the two sides.

File: signer/src/testing/transaction_signer.rs (hunk 2 above)
Comment (netrome) [reply signer/src/testing/transaction_signer.rs:356]: Makes sense. I can do it as a small follow-up if I find the time.

File: signer/src/transaction_coordinator.rs (hunk 1 above)
Comment (djordon) [reply signer/src/transaction_coordinator.rs]: All of those scenarios make sense to me. 

Okay, let me say the full context:
1. One of the signers is the coordinator and the coordinator is the only signer that can lead signing rounds for Stacks and BTC transactions.
2. Coordinators are the ones that initiate most communication, so if they receive a message it's probably a response from something that the coordinator initiated. Right now, everything else is dropped.
3. If a coordinator ceases to be the proper coordinator mid signing, they will miss signing requests from the new coordinator. This request might be a request that they would sign too (assuming their view of the blockchain has updated as well). The new coordinator now assumes the old one rejected the sign request.
4. The system still functions since only one signer is not fully "fulfilling its duties". Everything else should work just fine.

The part that I have the least clarity on is part (2). What communication does the coordinator drop during a signing round? How consequential is this information? I think the answer is that it they drop sign request messages only. Is that right?

Comment (netrome) [reply signer/src/transaction_coordinator.rs]: > If a coordinator ceases to be the proper coordinator mid signing, they will miss signing requests from the new coordinator. This request might be a request that they would sign too (assuming their view of the blockchain has updated as well). The new coordinator now assumes the old one rejected the sign request.

Just to clarify, the coordinator doesn't respond to signing requests - the signer does. So even if the transaction coordinator process hangs for a while, the same binary will have a transaction signer process (with the same keys as the coordinator process) running which will still be responsive to requests from other coordinators.

> The part that I have the least clarity on is part (2). What communication does the coordinator drop during a signing round? How consequential is this information? I think the answer is that it they drop sign request messages only. Is that right?

The coordinator isn't much of a listener and ignores most messages. When a coordinator is coordinating a signing round, it ignores all messages which are not wsts messages from other signers for the given chain tip. That is because these are the only messages providing the information the coordinator needs to produce a valid signature.

Comment (djordon) [reply signer/src/transaction_coordinator.rs]: Ahhh right, that makes sense.
